tiledb/sm/storage_manager/storage_manager.cc

```diff
@@ -1751,54 +1748,59 @@ Status StorageManager::load_array_schema(
 Status StorageManager::load_fragment_metadata(
     OpenArray* open_array,
     const EncryptionKey& encryption_key,
-    bool* in_cache) {
+    bool* in_cache,
+    uint64_t timestamp) {
   // Get all the fragment uris, sorted by timestamp
   std::vector<URI> fragment_uris;
   const URI& array_uri = open_array->array_uri();
   RETURN_NOT_OK(get_fragment_uris(array_uri, &fragment_uris));
-  sort_fragment_uris(&fragment_uris);
 
+  // Check if the array is empty
   if (fragment_uris.empty())
     return Status::Ok();
 
   *in_cache = false;
+  // Sort the URIs by timestamp
+  std::vector<std::pair<uint64_t, URI>> sorted_fragment_uris;
+  sort_fragment_uris(fragment_uris, &sorted_fragment_uris);
 
   // Load the metadata for each fragment, only if they are not already loaded
-  std::vector<FragmentMetadata*> fragment_metadata;
-  for (auto& uri : fragment_uris) {
-    auto metadata = open_array->fragment_metadata_get(uri);
-    if (metadata == nullptr) {
+  auto latest_timestamp = open_array->latest_timestamp();
+  for (auto& sf : sorted_fragment_uris) {
+    auto frag_timestamp = sf.first;
+    auto uri = sf.second;
+    if (frag_timestamp > latest_timestamp && frag_timestamp <= timestamp) {
       URI coords_uri =
           uri.join_path(constants::coords + constants::file_suffix);
       bool sparse;
       RETURN_NOT_OK(vfs_->is_file(coords_uri, &sparse));
-      metadata = new FragmentMetadata(open_array->array_schema(), !sparse, uri);
+      auto metadata = new FragmentMetadata(
+          open_array->array_schema(), !sparse, uri, frag_timestamp);
       bool metadata_in_cache;
       RETURN_NOT_OK_ELSE(
           load_fragment_metadata(metadata, encryption_key, &metadata_in_cache),
           delete metadata);
       *in_cache |= metadata_in_cache;
-      fragment_metadata.push_back(metadata);
+      open_array->push_back_fragment_metadata(metadata);
```

Comment:
Do we need to guarantee here that the vector in `open_array` stays sorted on timestamp when pushing to it? I am thinking there may be a multithreaded case where another thread can create a fragment with a lower-valued timestamp than the latest in the `OpenArray`, and when the array is reopened, that fragment gets appended to the back of this list rather than in its proper place.

Reply:
The order is vital for the reads (to know which fragment writes overwrite the previous ones in time). However, you are bringing up an important issue. We may have to insert and resort.